### PR TITLE
test: TLS SNI edge-case coverage for NameType, size, and trailing bytes

### LINKED ISSUE #52: test: add TLS SNI edge-case coverage for non-zero NameType, max length, and trailing bytes
## Context
Suggested by pr-test-analyzer review on PR #50 (S4). After #50 lands, the following SNI edge cases would still not be covered and are worth a second round of pin tests.

RFC 6066 §3 validation was done before filing (direct RFC fetch). Cases confirmed as real spec-level gaps are listed below; one case the reviewer suggested (duplicate SNI extension) was dropped because RFC 6066 is silent on duplicate extensions and testing it would pin library behavior rather than spec compliance.

### 1. Non-zero `NameType` values
RFC 6066 §3 defines:
```
enum {
    host_name(0), (255)
} NameType;
```
The `(255)` reserves 255 slots for future name types. The RFC body also says "other name types may be added in the future." A buggy or adversarial client could send a `ServerName` entry with `name_type` > 0 today. The current `build_client_hello_with_sni_list` helper hardcodes `sni_list_data.push(0x00)`, so this path is untestable from the existing fixture builder.

**What to pin:** what does `tls_parser` return when the `NameType` is not `host_name(0)`? Does `extract_sni` skip the entry, treat it as a hostname, or error? The analyzer's behavior should be explicitly documented in a pin test.

**Builder change needed:** extend `build_client_hello_with_sni_list` to take `&[(u8, &[u8])]` (type + bytes) instead of `&[&[u8]]`, OR add a lower-level raw-ServerNameList builder variant.

### 2. Maximum-length SNI at the u16 boundary
`HostName` in RFC 6066 is `opaque HostName<1..2^16-1>`, so the length field is u16 and max is 65,535 bytes. `MAX_BUF` in `TlsAnalyzer` is 65,536 (per-flow buffer cap), so a 65k-byte SNI would consume almost the entire buffer and leave nothing for subsequent records.

**What to pin:** does the analyzer parse a ClientHello with a 65,534-byte SNI hostname cleanly? Is `sni_counts` populated under the (huge) key? Does the per-flow buffer cap break anything?

### 3. Trailing bytes in ServerNameList
If the outer `sni_list_len` field lies — claiming more bytes than the sum of its ServerName entries — `tls_parser` may accept or reject the framing. A pin test documents which.

**What to pin:** the current tls_parser behavior + the analyzer's response. If tls_parser accepts with trailing garbage, does `extract_sni` still read the first entry correctly?

## Cases explicitly dropped

**Duplicate SNI extension in one ClientHello** — the reviewer suggested this, but RFC 6066 §3 does not specify duplicate-extension behavior (RFC 8446 TLS 1.3 §4.2 says "There MUST NOT be more than one extension of the same type in a given extension block" but not in RFC 6066 for TLS 1.2). Testing this would pin library behavior, not spec compliance — better left to a dedicated follow-up if there's real-world evidence it matters.

## Why not in #50
Each of these requires expanding the fixture builder (to take `(type, bytes)` tuples), and they're not what issue #50 asked for. #50's scope was the five cases the code-reviewer raised in S1 of PR #49's review; these are the S4 follow-ups from PR #50's own review.

## Reference
- PR #50 (issue #50 work)
- pr-test-analyzer finding S4
- RFC 6066 §3 (ServerNameList / NameType / HostName definitions)
- RFC 8446 §4.2 (for context on duplicate extension handling in TLS 1.3, not applicable to TLS 1.2)

### PR DESCRIPTION
**Pull request by Zious11**
## Summary

- Adds 5 tests and 2 new fixture builders closing all 3 edge cases from issue #52.
- Refactors existing `build_client_hello_with_sni_list` to delegate to a new typed variant.
- All pin tests empirically verified — behavior documented so tls_parser upgrades are caught.

### Edge cases covered
| Case | Tests | Pinned behavior |
|------|-------|----------------|
| Non-zero NameType | `test_non_zero_name_type_sni_entry`, `test_non_zero_name_type_with_valid_first_entry` | tls_parser includes unknown types; extract_sni ignores type field |
| Large/oversized SNI | `test_large_sni_near_record_payload_limit`, `test_oversized_sni_exceeds_record_payload_limit` | 16KB parses; >18,432 rejected at record layer |
| Trailing bytes | `test_trailing_bytes_in_server_name_list` | tls_parser tolerates trailing garbage; first entry readable |

### Note on u16 boundary (65K SNI)
The issue mentioned testing at the u16 max (65,535 bytes). This is unreachable through the production code path — `MAX_RECORD_PAYLOAD` (18,432 bytes) rejects the record before SNI parsing begins. The `test_oversized_sni_exceeds_record_payload_limit` test pins the actual effective boundary.

Closes #52

## Test plan
- [x] All 26 TLS analyzer tests pass
- [x] Full suite passes (176+ tests)
- [x] `cargo clippy --all-targets` clean
- [x] `cargo fmt --check` clean
- [x] Multi-agent PR review (test-analyzer, code-reviewer) — 1 issue fixed

### CODE REVIEW

File: tests/tls_analyzer_tests.rs [hunk 1]
```diff
@@ -816,3 +830,226 @@ fn test_multi_name_sni_list_only_first_entry_counted() {
...
+fn build_client_hello_with_raw_sni_ext(raw_sni_ext_data: &[u8], cipher_ids: &[u16]) -> Vec<u8> {
+    let mut extensions = Vec::new();
+
+    // SNI extension (type 0x0000) with caller-supplied raw data
+    let ext_data_len =
+        u16::try_from(raw_sni_ext_data.len()).expect("raw SNI ext data exceeds u16::MAX");
+    extensions.extend_from_slice(&[0x00, 0x00]); // extension type: server_name
+    extensions.extend_from_slice(&ext_data_len.to_be_bytes());
+    extensions.extend_from_slice(raw_sni_ext_data);
+
```
[the remark this answers is not in the document]
Comment (Zious11) [reply tests/tls_analyzer_tests.rs:853]: Valid point about duplication. However, the raw builder intentionally needs full control over the extension byte stream before framing — a shared helper would need to accept raw extension bytes anyway, making it essentially the same function with an extra indirection. Deferring this refactor since the two builders serve different purposes (structured vs. raw framing tests) and the duplication is contained to ~30 lines of boilerplate.

File: tests/tls_analyzer_tests.rs [hunk 2]
```diff
@@ -816,3 +830,226 @@ fn test_multi_name_sni_list_only_first_entry_counted() {
...
+    //   entry: NameType=0x00, name_len=12, "test.example"
+    let hostname = b"test.example";
+    let name_len = hostname.len() as u16;
+    let mut sni_list_data = Vec::new();
+    sni_list_data.push(0x00); // NameType = host_name
+    sni_list_data.extend_from_slice(&name_len.to_be_bytes());
+    sni_list_data.extend_from_slice(hostname);
+
+    // Lie about list length: claim 4 extra bytes of trailing garbage
+    let lying_list_len = (sni_list_data.len() + 4) as u16;
```
[the remark this answers is not in the document]
Comment (Zious11) [reply tests/tls_analyzer_tests.rs]: Fixed in 3f2d442 — both casts now use `u16::try_from(...).expect(...)` for consistency.

File: tests/tls_analyzer_tests.rs [hunk 3]
```diff
@@ -816,3 +830,228 @@ fn test_multi_name_sni_list_only_first_entry_counted() {
...
+fn test_oversized_sni_exceeds_record_payload_limit() {
+    // A SNI so large that the TLS record payload exceeds MAX_RECORD_PAYLOAD
+    // (18,432 bytes). The analyzer should reject the record at the record-layer
+    // boundary and increment parse_errors. The SNI is never reached.
+    let mut analyzer = TlsAnalyzer::new();
+    let fk = test_flow_key();
+
+    // 18,400 byte hostname + ~74 bytes overhead = ~18,474 > 18,432 limit
+    let huge_hostname = "B".repeat(18_400);
+    let record = build_client_hello(&huge_hostname, &[0x1301]);
```
[the remark this answers is not in the document]
Comment (Zious11) [reply tests/tls_analyzer_tests.rs]: Fixed in 45edd5b. The test now reads the record payload length from the header and asserts it exceeds MAX_RECORD_PAYLOAD *before* calling on_data — if the builder ever drifts and the payload drops below the cap, the test fails loudly with a clear precondition message instead of passing for the wrong reason.

File: tests/tls_analyzer_tests.rs [hunk 4]
```diff
@@ -816,3 +830,228 @@ fn test_multi_name_sni_list_only_first_entry_counted() {
...
+    // A large SNI hostname that fits within MAX_RECORD_PAYLOAD (18,432 bytes).
+    // The ClientHello overhead is ~74 bytes, so a 16,000-byte hostname produces
+    // a record payload of ~16,074 — well within the limit.
+    //
+    // This pins that a large-but-valid SNI parses cleanly and is counted.
+    let mut analyzer = TlsAnalyzer::new();
+    let fk = test_flow_key();
+
+    let large_hostname = "A".repeat(16_000);
+    let record = build_client_hello(&large_hostname, &[0x1301]);
```
[the remark this answers is not in the document]
Comment (Zious11) [reply tests/tls_analyzer_tests.rs:974]: Clarified in 45edd5b via comment. The large-SNI test intentionally pins size-handling behavior only — the wirerust TLS analyzer doesn't validate DNS label syntax, so a 16KB `"A".repeat(16_000)` is the simplest fixture that exercises the record-layer and SNI-parsing paths. Generating label-structured strings would add fixture complexity without changing what the test covers.

File: tests/tls_analyzer_tests.rs [hunk 5]
```diff
@@ -816,3 +830,242 @@ fn test_multi_name_sni_list_only_first_entry_counted() {
...
+    let huge_hostname = "B".repeat(18_400);
+    let record = build_client_hello(&huge_hostname, &[0x1301]);
+    assert!(
+        record.len() >= 5,
+        "fixture must produce a complete TLS record header"
+    );
+    let payload_len = u16::from_be_bytes([record[3], record[4]]) as usize;
+    assert!(
+        payload_len > 18_432,
+        "test precondition: record payload must exceed MAX_RECORD_PAYLOAD (got {payload_len})"
```
[the remark this answers is not in the document]
Comment (Zious11) [reply tests/tls_analyzer_tests.rs:1014]: Fixed in a30043b. Added a local `const MAX_RECORD_PAYLOAD: usize = 18_432;` with a sync note, mirroring the MAX_MAP_ENTRIES pattern in the capacity test.

File: tests/tls_analyzer_tests.rs [hunk 6]
```diff
@@ -32,19 +32,33 @@ fn build_client_hello_raw_sni(sni_bytes: &[u8], cipher_ids: &[u16]) -> Vec<u8> {
...
+
+/// Build a minimal TLS ClientHello with SNI entries where each entry has an explicit
+/// NameType byte. Used for testing non-zero NameType values (issue #52 case 1).
+///
+/// Each tuple is `(name_type, hostname_bytes)`. RFC 6066 §3 defines `host_name(0)`;
+/// the `(255)` in the enum reserves slots for future types. A non-zero NameType
+/// exercises the tls_parser's handling of unknown ServerName types.
+fn build_client_hello_with_typed_sni_list(
+    sni_entries: &[(u8, &[u8])],
+    cipher_ids: &[u16],
```
[the remark this answers is not in the document]
Comment (Zious11) [reply tests/tls_analyzer_tests.rs:49]: Fixed in a30043b. Doc now says `(name_type, name_bytes)` and clarifies that the bytes are an ASCII hostname only for host_name(0); for other NameTypes they are the opaque ServerName payload for that type.

File: tests/tls_analyzer_tests.rs [hunk 7]
```diff
@@ -32,19 +32,35 @@ fn build_client_hello_raw_sni(sni_bytes: &[u8], cipher_ids: &[u16]) -> Vec<u8> {
...
+    // The ServerNameList contents: for each entry, 1 byte NameType + 2 byte name length + name bytes.
     // Use checked conversions so a future test passing an out-of-range SNI (e.g. a
     // ~65,532-byte single hostname — u16::MAX minus 3 bytes of ServerName overhead
     // per entry — requested in issue #52) panics with a clear message rather than
     // silently wrapping through `as u16` and producing a malformed ClientHello.
     let mut sni_list_data = Vec::new();
-    for entry in sni_entries {
+    for (name_type, entry) in sni_entries {
         let name_len = u16::try_from(entry.len())
             .expect("SNI hostname entry exceeds u16::MAX; can't encode as RFC 6066 HostName");
```
[the remark this answers is not in the document]
Comment (Zious11) [reply tests/tls_analyzer_tests.rs]: Fixed in fe20fd7. The message now reads "SNI entry payload exceeds u16::MAX; can't encode as ServerName" — type-agnostic and accurate for non-zero NameType entries.
